.github/workflows/ci.yml

```diff
@@ -45,21 +84,30 @@ jobs:
           - server_type: dotnet
             server_image: mcr.microsoft.com/dotnet/sdk:6.0
     steps:
-      - uses: actions/checkout@v3
+      - id: if_running
+        run: |
+          prefix="${{ matrix.target.sample }}/server/${{ matrix.runtime.server_type }}"
+          if [[ "${{ github.event.pull_request.head.ref || github.ref }}" = "refs/heads/main" ]] || \
+            [[ $(echo "${{ needs.changed_files.outputs.files }}" | grep -E "^${prefix}") ]]
+          then echo 'running=true' >> $GITHUB_OUTPUT
+          fi
+
+      - uses: actions/checkout@v4
+        if: ${{ steps.if_running.outputs.running }}
+        with:
+          ref: ${{ github.event.after || github.event.pull_request.head.sha }}
 
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
```

Comment:
Updated `actions/checkout` (from v3 to v4) and `actions/upload-artifact` (from v2 to v4) since the older versions are deprecated.